File: src/utils/morphEngine.ts
import { PatchParameters, ImageMetrics, PhysicalModelParameters, ImageEffectsState, TemporalEraInfo, EraYear } from '../types/synth';
import { hexToRgb, rgbToHex } from './colorBlending';

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

function lerpRound(a: number, b: number, t: number): number {
  return Math.round(lerp(a, b, t));
}

function lerpColorHex(hexA: string, hexB: string, t: number): string {
  const [rA, gA, bA] = hexToRgb(hexA || '#00F2FE');
  const [rB, gB, bB] = hexToRgb(hexB || '#4FACFE');
  const r = Math.round(lerp(rA, rB, t));
  const g = Math.round(lerp(gA, gB, t));
  const b = Math.round(lerp(bA, bB, t));
  return rgbToHex(r, g, b);
}

function interpolateTemporalEraInfo(
  eraA?: TemporalEraInfo,
  eraB?: TemporalEraInfo,
  t: number = 0.5
): TemporalEraInfo | undefined {
  if (!eraA && !eraB) return undefined;
  if (!eraA) return eraB;
  if (!eraB) return eraA;

  const clampT = Math.max(0, Math.min(1, t));
  const eraVal = Number(lerp(eraA.eraVal, eraB.eraVal, clampT).toFixed(3));

  const w1800s = Number(lerp(eraA.weights.era1800s, eraB.weights.era1800s, clampT).toFixed(3));
  const w1960s = Number(lerp(eraA.weights.era1960s, eraB.weights.era1960s, clampT).toFixed(3));
  const w1970s = Number(lerp(eraA.weights.era1970s, eraB.weights.era1970s, clampT).toFixed(3));
  const w1980s = Number(lerp(eraA.weights.era1980s, eraB.weights.era1980s, clampT).toFixed(3));
  const w2000s = Number(lerp(eraA.weights.era2000s, eraB.weights.era2000s, clampT).toFixed(3));

  let eraYear: EraYear = '1970s';
  let label = '1970s Discrete Analog';
  let architecture = 'Moog Ladder Monophonic & Rhodes';
  let description = 'Warm, fat discrete analog synths, ARP 2600 modular screams & electric pianos';

  if (eraVal < 0.16) {
    eraYear = '1800s & Prior';
    label = '1800s Physical & Cathedral';
    architecture = 'Harpsichord, Gut Lute & Cathedral Organ';
    description = 'Plucked harpsichords, gut-string lutes & cathedral pipe organ models with church acoustic impulse responses';
  } else if (eraVal < 0.38) {
    eraYear = '1960s';
    label = '1960s Drawbar & Combo Organ';
    architecture = 'Hammond/Vox Drawbar Organ & Spring Tape';
    description = 'Psychedelic drawbar tonewheel organs, tape-sands & spring-reverbed transistor combo organs';
  } else if (eraVal < 0.62) {
    eraYear = '1970s';
    label = '1970s Discrete Analog';
    architecture = 'Moog Ladder & Fender Rhodes';
    description = 'Warm, fat discrete analog synths, ARP 2600 modular screams & electric pianos';
  } else if (eraVal < 0.84) {
    eraYear = '1980s';
    label = '1980s Polyphonic OB-X & DX7';
    architecture = 'Oberheim OB-X & DX7 FM Bells';
    description = 'Sweeping polyphonic powerhouses, Roland Juno chorus pads & DX7 digital FM bells';
  } else {
    eraYear = '1990s-2000s+';
    label = '2000s Digital Wavetable & Sampler';
    architecture = 'Digital Wavetables & Modern Hybrid';
    description = 'Crisp digital wavetables, lo-fi sampler grit & high-fidelity modern hybrid pads';
  }

  return {
    eraVal,
    eraYear,
    label,
    architecture,
    description,
    weights: {
      era1800s: w1800s,
      era1960s: w1960s,
      era1970s: w1970s,
      era1980s: w1980s,
      era2000s: w2000s,
    },
  };
}

export function interpolatePatchParameters(
  patchA: PatchParameters,
  patchB: PatchParameters,
  t: number
): PatchParameters {
  const clampT = Math.max(0, Math.min(1, t));

  // Custom LFO profile interpolation
  let customLfoProfile: number[] | undefined;
  if (patchA.customLfoProfile && patchB.customLfoProfile) {
    const len = Math.min(patchA.customLfoProfile.length, patchB.customLfoProfile.length);
    customLfoProfile = [];
    for (let i = 0; i < len; i++) {
      customLfoProfile.push(
        Number(lerp(patchA.customLfoProfile[i], patchB.customLfoProfile[i], clampT).toFixed(3))
      );
    }
  } else {
    customLfoProfile = patchA.customLfoProfile || patchB.customLfoProfile;
  }

  // Physical Model interpolation
  let physicalModel: PhysicalModelParameters | undefined;
  if (patchA.physicalModel && patchB.physicalModel) {
    const pmA = patchA.physicalModel;
    const pmB = patchB.physicalModel;

    const freqsA = pmA.bodyResonanceFreqs || [180, 520, 1400];
    const freqsB = pmB.bodyResonanceFreqs || [180, 520, 1400];

    physicalModel = {
      materialType: clampT < 0.5 ? pmA.materialType : pmB.materialType,
      bodyResonanceFreqs: [
        Math.round(lerp(freqsA[0], freqsB[0], clampT)),
        Math.round(lerp(freqsA[1], freqsB[1], clampT)),
        Math.round(lerp(freqsA[2], freqsB[2], clampT)),
      ],
      bodyDamping: Number(lerp(pmA.bodyDamping, pmB.bodyDamping, clampT).toFixed(2)),
      acousticWeight: Number(lerp(pmA.acousticWeight, pmB.acousticWeight, clampT).toFixed(2)),
      tapeFlutterSpeed: Number(lerp(pmA.tapeFlutterSpeed, pmB.tapeFlutterSpeed, clampT).toFixed(2)),
      tapeFlutterDepth: Number(lerp(pmA.tapeFlutterDepth, pmB.tapeFlutterDepth, clampT).toFixed(2)),
      analogSaturationWarmth: Number(
        lerp(pmA.analogSaturationWarmth, pmB.analogSaturationWarmth, clampT).toFixed(2)
      ),
      transientProfile: {
        type: clampT < 0.5 ? pmA.transientProfile.type : pmB.transientProfile.type,
        hardness: Number(lerp(pmA.transientProfile.hardness, pmB.transientProfile.hardness, clampT).toFixed(2)),
        noiseBurst: Number(
          lerp(pmA.transientProfile.noiseBurst, pmB.transientProfile.noiseBurst, clampT).toFixed(2)
        ),
      },
      spatialReflection: {
        roomSizeSeconds: Number(
          lerp(pmA.spatialReflection.roomSizeSeconds, pmB.spatialReflection.roomSizeSeconds, clampT).toFixed(2)
        ),
        absorption: Number(
          lerp(pmA.spatialReflection.absorption, pmB.spatialReflection.absorption, clampT).toFixed(2)
        ),
        diffusion: Number(
          lerp(pmA.spatialReflection.diffusion, pmB.spatialReflection.diffusion, clampT).toFixed(2)
        ),
        luminanceDepth: Number(
          lerp(pmA.spatialReflection.luminanceDepth, pmB.spatialReflection.luminanceDepth, clampT).toFixed(2)
        ),
      },
    };
  } else {
    physicalModel = patchA.physicalModel || patchB.physicalModel;
  }

  // Effects interpolation
  let effects: ImageEffectsState | undefined;
  if (patchA.effects && patchB.effects) {
    const effA = patchA.effects;
    const effB = patchB.effects;
    const keys: (keyof ImageEffectsState)[] = ['delay', 'reverb', 'chorus', 'phaser', 'flanger', 'distortion'];
    effects = { ...effA };
    keys.forEach((k) => {
      effects![k] = {
        ...effA[k],
        enabled: clampT < 0.5 ? effA[k].enabled : effB[k].enabled,
        intensity: Number(lerp(effA[k].intensity, effB[k].intensity, clampT).toFixed(2)),
      };
    });
  } else {
    effects = patchA.effects || patchB.effects;
  }

  return {
    engineType: clampT < 0.5 ? patchA.engineType : patchB.engineType,
    cutoffOffset: Math.round(lerp(patchA.cutoffOffset, patchB.cutoffOffset, clampT)),
    resonance: Number(lerp(patchA.resonance, patchB.resonance, clampT).toFixed(2)),
    lfoRate: Number(lerp(patchA.lfoRate, patchB.lfoRate, clampT).toFixed(2)),
    lfoDepth: Number(lerp(patchA.lfoDepth, patchB.lfoDepth, clampT).toFixed(2)),
    masterVolume: Number(lerp(patchA.masterVolume, patchB.masterVolume, clampT).toFixed(1)),
    limiterEnabled: clampT < 0.5 ? patchA.limiterEnabled : patchB.limiterEnabled,
    threshold: Number(lerp(patchA.threshold, patchB.threshold, clampT).toFixed(1)),
    attackTime: Number(lerp(patchA.attackTime, patchB.attackTime, clampT).toFixed(3)),
    decayTime: Number(lerp(patchA.decayTime, patchB.decayTime, clampT).toFixed(3)),
    sustainLevel: Number(lerp(patchA.sustainLevel, patchB.sustainLevel, clampT).toFixed(2)),
    releaseTime: Number(lerp(patchA.releaseTime, patchB.releaseTime, clampT).toFixed(3)),
    detuneCents: Number(lerp(patchA.detuneCents, patchB.detuneCents, clampT).toFixed(1)),
    unisonVoices: lerpRound(patchA.unisonVoices, patchB.unisonVoices, clampT),
    fmRatio: Number(lerp(patchA.fmRatio, patchB.fmRatio, clampT).toFixed(2)),
    fmAmount: Number(lerp(patchA.fmAmount, patchB.fmAmount, clampT).toFixed(2)),
    subOscLevel: Number(lerp(patchA.subOscLevel, patchB.subOscLevel, clampT).toFixed(2)),
    reverbMix: Number(lerp(patchA.reverbMix, patchB.reverbMix, clampT).toFixed(2)),
    delayMix: Number(lerp(patchA.delayMix, patchB.delayMix, clampT).toFixed(2)),
    customLfoProfile,
    physicalModel,
    effects,
    temporalEraVal: Number(
      lerp(
        patchA.temporalEraVal ?? patchA.temporalEra?.eraVal ?? 0.5,
        patchB.temporalEraVal ?? patchB.temporalEra?.eraVal ?? 0.5,
        clampT
      ).toFixed(3)
    ),
    temporalEra: interpolateTemporalEraInfo(patchA.temporalEra, patchB.temporalEra, clampT),
    opticalFocusDepth: Number(lerp(patchA.opticalFocusDepth ?? 0.5, patchB.opticalFocusDepth ?? 0.5, clampT).toFixed(3)),
    gridSymmetryDensity: Number(lerp(patchA.gridSymmetryDensity ?? 0.5, patchB.gridSymmetryDensity ?? 0.5, clampT).toFixed(3)),
    lightAzimuthAngle: Math.round(lerp(patchA.lightAzimuthAngle ?? 180, patchB.lightAzimuthAngle ?? 180, clampT)),
    lightElevationAngle: Math.round(lerp(patchA.lightElevationAngle ?? 45, patchB.lightElevationAngle ?? 45, clampT)),
    chromaticClash: Number(lerp(patchA.chromaticClash ?? 0.2, patchB.chromaticClash ?? 0.2, clampT).toFixed(3)),
    semanticDensityWeight: Number(lerp(patchA.semanticDensityWeight ?? 0.5, patchB.semanticDensityWeight ?? 0.5, clampT).toFixed(3)),
  };
}

export function interpolateImageMetrics(
  metricsA: ImageMetrics,
  metricsB: ImageMetrics,
  t: number
): ImageMetrics {
  const clampT = Math.max(0, Math.min(1, t));

  // Blend colorPalette array
  let colorPalette: string[] | undefined;
  if (metricsA.colorPalette && metricsB.colorPalette) {
    const len = Math.max(metricsA.colorPalette.length, metricsB.colorPalette.length);
    colorPalette = [];
    for (let i = 0; i < len; i++) {
      const colA = metricsA.colorPalette[i % metricsA.colorPalette.length];
      const colB = metricsB.colorPalette[i % metricsB.colorPalette.length];
      colorPalette.push(lerpColorHex(colA, colB, clampT));
    }
  } else {
    colorPalette = metricsA.colorPalette || metricsB.colorPalette;
  }

  return {
    hash64: clampT < 0.5 ? metricsA.hash64 : metricsB.hash64,
    seedNumber: clampT < 0.5 ? metricsA.seedNumber : metricsB.seedNumber,
    timbreDna: Number(lerp(metricsA.timbreDna, metricsB.timbreDna, clampT).toFixed(3)),
    brightness: Number(lerp(metricsA.brightness, metricsB.brightness, clampT).toFixed(3)),
    saturation: Number(lerp(metricsA.saturation, metricsB.saturation, clampT).toFixed(3)),
    complexity: Number(lerp(metricsA.complexity, metricsB.complexity, clampT).toFixed(3)),
    dominantHue: Math.round(lerp(metricsA.dominantHue, metricsB.dominantHue, clampT)),
    archetype: clampT < 0.5 ? metricsA.archetype : metricsB.archetype,
    colorPalette,
    detectedStyle: clampT < 0.5 ? metricsA.detectedStyle : metricsB.detectedStyle,
    temporalEra: interpolateTemporalEraInfo(metricsA.temporalEra, metricsB.temporalEra, clampT),
    backgroundFeatures: (metricsA.backgroundFeatures && metricsB.backgroundFeatures) ? {
      opticalFocusDepth: Number(lerp(metricsA.backgroundFeatures.opticalFocusDepth, metricsB.backgroundFeatures.opticalFocusDepth, clampT).toFixed(3)),
      gridSymmetryDensity: Number(lerp(metricsA.backgroundFeatures.gridSymmetryDensity, metricsB.backgroundFeatures.gridSymmetryDensity, clampT).toFixed(3)),
      lightAzimuthAngle: Math.round(lerp(metricsA.backgroundFeatures.lightAzimuthAngle, metricsB.backgroundFeatures.lightAzimuthAngle, clampT)),
      lightElevationAngle: Math.round(lerp(metricsA.backgroundFeatures.lightElevationAngle, metricsB.backgroundFeatures.lightElevationAngle, clampT)),
      chromaticClash: Number(lerp(metricsA.backgroundFeatures.chromaticClash, metricsB.backgroundFeatures.chromaticClash, clampT).toFixed(3)),
      semanticDensityWeight: Number(lerp(metricsA.backgroundFeatures.semanticDensityWeight, metricsB.backgroundFeatures.semanticDensityWeight, clampT).toFixed(3)),
      detectedMaterial: clampT < 0.5 ? metricsA.backgroundFeatures.detectedMaterial : metricsB.backgroundFeatures.detectedMaterial,
    } : (metricsA.backgroundFeatures || metricsB.backgroundFeatures),
  };
}

File: Source/MorphEngine.cpp
#include "MorphEngine.h"

namespace photosynth
{
    float MorphEngine::lerp(float a, float b, float t)
    {
        return a + (b - a) * juce::jlimit(0.0f, 1.0f, t);
    }

    TemporalEraInfo MorphEngine::interpolateTemporalEra(const TemporalEraInfo& a,
                                                        const TemporalEraInfo& b,
                                                        float t)
    {
        TemporalEraInfo out;
        out.eraVal = lerp(a.eraVal, b.eraVal, t);
        out.weights.era1800s = lerp(a.weights.era1800s, b.weights.era1800s, t);
        out.weights.era1960s = lerp(a.weights.era1960s, b.weights.era1960s, t);
        out.weights.era1970s = lerp(a.weights.era1970s, b.weights.era1970s, t);
        out.weights.era1980s = lerp(a.weights.era1980s, b.weights.era1980s, t);
        out.weights.era2000s = lerp(a.weights.era2000s, b.weights.era2000s, t);

        if (out.eraVal < 0.16f)
        {
            out.eraYear = EraYear::era1800s;
            out.label = "1800s Physical & Cathedral";
            out.architecture = "Harpsichord, Gut Lute & Cathedral Organ";
        }
        else if (out.eraVal < 0.38f)
        {
            out.eraYear = EraYear::era1960s;
            out.label = "1960s Drawbar & Combo Organ";
            out.architecture = "Hammond/Vox Drawbar Organ & Spring Tape";
        }
        else if (out.eraVal < 0.62f)
        {
            out.eraYear = EraYear::era1970s;
            out.label = "1970s Discrete Analog";
            out.architecture = "Moog Ladder & Fender Rhodes";
        }
        else if (out.eraVal < 0.84f)
        {
            out.eraYear = EraYear::era1980s;
            out.label = "1980s Polyphonic OB-X & DX7";
            out.architecture = "Oberheim OB-X & DX7 FM Bells";
        }
        else
        {
            out.eraYear = EraYear::era2000s;
            out.label = "2000s Digital Wavetable & Sampler";
            out.architecture = "Digital Wavetables & Modern Hybrid";
        }
        return out;
    }

    PatchParameters MorphEngine::interpolatePatchParameters(const PatchParameters& a,
                                                            const PatchParameters& b,
                                                            float t)
    {
        PatchParameters out;
        const float x = juce::jlimit(0.0f, 1.0f, t);

        out.engineType = x < 0.5f ? a.engineType : b.engineType;
        out.cutoffOffset = std::round(lerp(a.cutoffOffset, b.cutoffOffset, x));
        out.resonance = lerp(a.resonance, b.resonance, x);
        out.lfoRate = lerp(a.lfoRate, b.lfoRate, x);
        out.lfoDepth = lerp(a.lfoDepth, b.lfoDepth, x);

        out.masterVolume = lerp(a.masterVolume, b.masterVolume, x);
        out.attackTime = lerp(a.attackTime, b.attackTime, x);
        out.decayTime = lerp(a.decayTime, b.decayTime, x);
        out.sustainLevel = lerp(a.sustainLevel, b.sustainLevel, x);
        out.releaseTime = lerp(a.releaseTime, b.releaseTime, x);
        out.limiterEnabled = x < 0.5f ? a.limiterEnabled : b.limiterEnabled;
        out.threshold = lerp(a.threshold, b.threshold, x);

        out.detuneCents = lerp(a.detuneCents, b.detuneCents, x);
        out.unisonVoices = (int) std::round(lerp((float) a.unisonVoices, (float) b.unisonVoices, x));
        out.fmRatio = lerp(a.fmRatio, b.fmRatio, x);
        out.fmAmount = lerp(a.fmAmount, b.fmAmount, x);
        out.subOscLevel = lerp(a.subOscLevel, b.subOscLevel, x);
        out.reverbMix = lerp(a.reverbMix, b.reverbMix, x);
        out.delayMix = lerp(a.delayMix, b.delayMix, x);

        out.customLfoProfile = a.customLfoProfile;
        if (!a.customLfoProfile.empty() && !b.customLfoProfile.empty())
        {
            const size_t len = juce::jmin(a.customLfoProfile.size(), b.customLfoProfile.size());
            out.customLfoProfile.resize(len);
            for (size_t i = 0; i < len; ++i)
                out.customLfoProfile[i] = lerp(a.customLfoProfile[i], b.customLfoProfile[i], x);
        }

        out.physicalModel = x < 0.5f ? a.physicalModel : b.physicalModel;
        for (int i = 0; i < 3; ++i)
            out.physicalModel.bodyResonanceFreqs[(size_t)i] = lerp(a.physicalModel.bodyResonanceFreqs[(size_t)i], b.physicalModel.bodyResonanceFreqs[(size_t)i], x);
        out.physicalModel.bodyDamping = lerp(a.physicalModel.bodyDamping, b.physicalModel.bodyDamping, x);
        out.physicalModel.acousticWeight = lerp(a.physicalModel.acousticWeight, b.physicalModel.acousticWeight, x);
        out.physicalModel.tapeFlutterSpeed = lerp(a.physicalModel.tapeFlutterSpeed, b.physicalModel.tapeFlutterSpeed, x);
        out.physicalModel.tapeFlutterDepth = lerp(a.physicalModel.tapeFlutterDepth, b.physicalModel.tapeFlutterDepth, x);
        out.physicalModel.analogSaturationWarmth = lerp(a.physicalModel.analogSaturationWarmth, b.physicalModel.analogSaturationWarmth, x);
        out.physicalModel.transientProfile.hardness = lerp(a.physicalModel.transientProfile.hardness, b.physicalModel.transientProfile.hardness, x);
        out.physicalModel.transientProfile.noiseBurst = lerp(a.physicalModel.transientProfile.noiseBurst, b.physicalModel.transientProfile.noiseBurst, x);

        auto blendEffect = [x](const EffectConfig& ea, const EffectConfig& eb)
        {
            EffectConfig e;
            e.enabled = x < 0.5f ? ea.enabled : eb.enabled;
            e.intensity = ea.intensity + (eb.intensity - ea.intensity) * x;
            e.param1 = ea.param1 + (eb.param1 - ea.param1) * x;
            e.param2 = ea.param2 + (eb.param2 - ea.param2) * x;
            return e;
        };

        out.effects.delay = blendEffect(a.effects.delay, b.effects.delay);
        out.effects.reverb = blendEffect(a.effects.reverb, b.effects.reverb);
        out.effects.chorus = blendEffect(a.effects.chorus, b.effects.chorus);
        out.effects.phaser = blendEffect(a.effects.phaser, b.effects.phaser);
        out.effects.flanger = blendEffect(a.effects.flanger, b.effects.flanger);
        out.effects.distortion = blendEffect(a.effects.distortion, b.effects.distortion);

        out.temporalEraVal = lerp(a.temporalEraVal, b.temporalEraVal, x);
        out.temporalEra = interpolateTemporalEra(a.temporalEra, b.temporalEra, x);

        out.opticalFocusDepth = lerp(a.opticalFocusDepth, b.opticalFocusDepth, x);
        out.gridSymmetryDensity = lerp(a.gridSymmetryDensity, b.gridSymmetryDensity, x);
        out.lightAzimuthAngle = std::round(lerp(a.lightAzimuthAngle, b.lightAzimuthAngle, x));
        out.lightElevationAngle = std::round(lerp(a.lightElevationAngle, b.lightElevationAngle, x));
        out.chromaticClash = lerp(a.chromaticClash, b.chromaticClash, x);
        out.semanticDensityWeight = lerp(a.semanticDensityWeight, b.semanticDensityWeight, x);

        return out;
    }

    ImageMetrics MorphEngine::interpolateImageMetrics(const ImageMetrics& a,
                                                      const ImageMetrics& b,
                                                      float t)
    {
        ImageMetrics out;
        const float x = juce::jlimit(0.0f, 1.0f, t);
        out.hash64 = x < 0.5f ? a.hash64 : b.hash64;
        out.seedNumber = x < 0.5f ? a.seedNumber : b.seedNumber;
        out.timbreDna = lerp(a.timbreDna, b.timbreDna, x);
        out.brightness = lerp(a.brightness, b.brightness, x);
        out.saturation = lerp(a.saturation, b.saturation, x);
        out.complexity = lerp(a.complexity, b.complexity, x);
        out.dominantHue = std::round(lerp(a.dominantHue, b.dominantHue, x));
        out.archetype = x < 0.5f ? a.archetype : b.archetype;
        out.temporalEra = interpolateTemporalEra(a.temporalEra, b.temporalEra, x);
        out.backgroundFeatures.opticalFocusDepth = lerp(a.backgroundFeatures.opticalFocusDepth, b.backgroundFeatures.opticalFocusDepth, x);
        out.backgroundFeatures.gridSymmetryDensity = lerp(a.backgroundFeatures.gridSymmetryDensity, b.backgroundFeatures.gridSymmetryDensity, x);
        out.backgroundFeatures.lightAzimuthAngle = std::round(lerp(a.backgroundFeatures.lightAzimuthAngle, b.backgroundFeatures.lightAzimuthAngle, x));
        out.backgroundFeatures.lightElevationAngle = std::round(lerp(a.backgroundFeatures.lightElevationAngle, b.backgroundFeatures.lightElevationAngle, x));
        out.backgroundFeatures.chromaticClash = lerp(a.backgroundFeatures.chromaticClash, b.backgroundFeatures.chromaticClash, x);
        out.backgroundFeatures.semanticDensityWeight = lerp(a.backgroundFeatures.semanticDensityWeight, b.backgroundFeatures.semanticDensityWeight, x);
        out.backgroundFeatures.detectedMaterial = x < 0.5f ? a.backgroundFeatures.detectedMaterial : b.backgroundFeatures.detectedMaterial;
        return out;
    }
}
